pkg/gateway/filter/handler/grpc/handler.go

```diff
@@ -0,0 +1,587 @@
+	loc, err := s.gateway.Insert(ctx, vec, s.copts...)
+	if err != nil {
+		err = errors.Wrapf(err, "Insert API (do multiple) failed to Insert uuid = %s\tmeta = %s\t info = %#v", uuid, meta, info.Get())
+		log.Debug(err)
+		if span != nil {
+			span.SetStatus(trace.StatusCodeInternal(err.Error()))
+		}
+		return nil, status.WrapWithInternal(fmt.Sprintf("Insert API failed to Execute DoMulti error = %s", err.Error()), err, info.Get())
+	}
+	return loc, nil
+}
+
+func (s *server) StreamInsert(stream vald.Vald_StreamInsertServer) error {
+	ctx, span := trace.StartSpan(stream.Context(), apiName+".StreamInsert")
+	defer func() {
+		if span != nil {
+			span.End()
+		}
+	}()
+	return grpc.BidirectionalStream(ctx, stream, s.streamConcurrency,
+		func() interface{} { return new(payload.Object_Vector) },
+		func(ctx context.Context, data interface{}) (interface{}, error) {
+			return s.Insert(ctx, data.(*payload.Object_Vector))
+		})
+}
+
+func (s *server) MultiInsert(ctx context.Context, vecs *payload.Object_Vectors) (res *payload.Object_Locations, err error) {
+	ctx, span := trace.StartSpan(ctx, apiName+".MultiInsert")
+	defer func() {
+		if span != nil {
+			span.End()
+		}
+	}()
+	metaMap := make(map[string]string, len(vecs.GetVectors()))
+	metas := make([]string, 0, len(vecs.GetVectors()))
+	for i, vec := range vecs.GetVectors() {
+		uuid := fuid.String()
+		meta := vec.GetId()
+		metaMap[uuid] = meta
+		metas = append(metas, meta)
+		vecs.Vectors[i].Id = uuid
+	}
+
+	for _, meta := range metas {
+		exists, err := s.metadata.Exists(ctx, meta)
+		if err != nil {
+			log.Debug(err)
+			if span != nil {
+				span.SetStatus(trace.StatusCodeInternal(err.Error()))
+			}
+			return nil, status.WrapWithInternal(
+				fmt.Sprintf("MultiInsert API couldn't check metadata exists or not metas = %v", metas), err, info.Get())
+		}
+		if exists {
+			if span != nil {
+				span.SetStatus(trace.StatusCodeAlreadyExists(err.Error()))
+			}
+			return nil, status.WrapWithAlreadyExists(
+				fmt.Sprintf("MultiInsert API failed metadata already exists meta = %s", meta), err, info.Get())
+		}
+	}
+
+	res, err = s.gateway.MultiInsert(ctx, vecs, s.copts...)
+	if err != nil {
+		log.Debug(err)
+		if span != nil {
+			span.SetStatus(trace.StatusCodeInternal(err.Error()))
+		}
+		return nil, status.WrapWithInternal("got error from MultiInsert API", err, info.Get())
+	}
+
+	err = s.metadata.SetUUIDandMetas(ctx, metaMap)
+	if err != nil {
+		if span != nil {
+			span.SetStatus(trace.StatusCodeInternal(err.Error()))
+		}
+		return nil, status.WrapWithInternal(fmt.Sprintf("MultiInsert API failed SetUUIDandMetas %#v", metaMap), err, info.Get())
+	}
+
+	return res, nil
+}
+
+func (s *server) Update(ctx context.Context, vec *payload.Object_Vector) (res *payload.Object_Location, err error) {
+	ctx, span := trace.StartSpan(ctx, apiName+".Update")
+	defer func() {
+		if span != nil {
+			span.End()
+		}
+	}()
+	meta := vec.GetId()
+	uuid, err := s.metadata.GetUUID(ctx, meta)
```

Comment:
Please teach me about the following lines. :bow:
From my own understanding, I thought that filter does not connect to the Vald Meta, but only proxies to the underlying gateway (Vald MetaGateway). 🤔

Reply:
filter gateway is another task please ignore them.